pkg/server/plugin/nodeattestor/k8s/k8s.go

```diff
@@ -0,0 +1,210 @@
+
+	if leaf.Subject.CommonName != "" {
+		selectors = append(selectors, &common.Selector{
+			Type: "k8s", Value: "subject:cn:" + leaf.Subject.CommonName,
+		})
+	}
+
+	// Used to avoid duplicating selectors.
+	fingerprints := map[string]*x509.Certificate{}
+	for _, chain := range chains {
+		// Iterate over all the certs in the chain (skip leaf at the 0 index)
+		for _, cert := range chain[1:] {
+			fp := x509pop.Fingerprint(cert)
+			// If the same fingerprint is generated, continue with the next certificate, because
+			// a selector should have been already created for it.
+			if _, ok := fingerprints[fp]; ok {
+				continue
+			}
+			fingerprints[fp] = cert
+
+			selectors = append(selectors, &common.Selector{
+				Type: "k8s", Value: "ca:fingerprint:" + fp,
+			})
+		}
+	}
+
+	return selectors
+}
+
+func New() *K8sPlugin {
+	return &K8sPlugin{}
+}
+
+func (p *K8sPlugin) Attest(stream nodeattestor.Attest_PluginStream) error {
+	req, err := stream.Recv()
+	if err != nil {
+		return err
+	}
+
+	c := p.getConfiguration()
+	if c == nil {
+		return newError("not configured")
+	}
+
+	if dataType := req.AttestationData.Type; dataType != pluginName {
+		return newError("unexpected attestation data type %q", dataType)
+	}
+
+	attestationData := new(x509pop.AttestationData)
+	if err := json.Unmarshal(req.AttestationData.Data, attestationData); err != nil {
+		return newError("failed to unmarshal data: %v", err)
+	}
+
+	// build up leaf certificate and list of intermediates
+	if len(attestationData.Certificates) == 0 {
+		return newError("no certificate to attest")
+	}
+	leaf, err := x509.ParseCertificate(attestationData.Certificates[0])
+	if err != nil {
+		return newError("unable to parse leaf certificate: %v", err)
+	}
+	intermediates := x509.NewCertPool()
+	for i, intermediateBytes := range attestationData.Certificates[1:] {
+		intermediate, err := x509.ParseCertificate(intermediateBytes)
+		if err != nil {
+			return newError("unable to parse intermediate certificate %d: %v", i, err)
+		}
+		intermediates.AddCert(intermediate)
+	}
+
+	// verify the chain of trust
+	chains, err := leaf.Verify(x509.VerifyOptions{
+		Intermediates: intermediates,
+		Roots:         c.trustBundle,
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+	})
+	if err != nil {
+		return newError("certificate verification failed: %v", err)
+	}
+
+	// now that the leaf certificate is trusted, issue a challenge to the node
+	// to prove possession of the private key.
+	challenge, err := x509pop.GenerateChallenge(leaf)
+	if err != nil {
+		return fmt.Errorf("unable to generate challenge: %v", err)
+	}
+
+	challengeBytes, err := json.Marshal(challenge)
+	if err != nil {
+		return fmt.Errorf("unable to marshal challenge: %v", err)
+	}
+
+	if err := stream.Send(&nodeattestor.AttestResponse{
+		Challenge: challengeBytes,
+	}); err != nil {
+		return err
+	}
+
+	// receive and validate the challenge response
+	responseReq, err := stream.Recv()
```

Comment:
Hmmm... I am curious if this can be abused by a client that never sends the challenge response... or triggered by clients which crash before being able to do so? Perhaps this `Recv` should be guarded with a select block on context cancellation?

@azdagron does the streaming shim have a default timeout on the context here? Is this a problem elsewhere?

Reply:
The `Recv()` call should return with an error in the case of context cancellation. You bring up a good point though. We should probably define timeouts for all plugin related calls. I don't think there is anything specific to do in the plugin code.